// [[Reason: undefined]] ////

//// [tests/cases/conformance/ambient/ambientErrors.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -39,8 +39,9 @@
 
 ambientErrors.ts(2,17): error TS1039: Initializers are not allowed in ambient contexts.
 ambientErrors.ts(17,22): error TS2371: A parameter initializer is only allowed in a function or constructor implementation.
 ambientErrors.ts(20,30): error TS1183: An implementation cannot be declared in ambient contexts.
+ambientErrors.ts(29,5): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 ambientErrors.ts(29,9): error TS1066: In ambient enum declarations member initializer must be constant expression.
 ambientErrors.ts(34,13): error TS1039: Initializers are not allowed in ambient contexts.
 ambientErrors.ts(35,25): error TS1183: An implementation cannot be declared in ambient contexts.
 ambientErrors.ts(37,20): error TS1039: Initializers are not allowed in ambient contexts.
@@ -52,9 +53,9 @@
 ambientErrors.ts(51,16): error TS2436: Ambient module declaration cannot specify relative module name.
 ambientErrors.ts(57,5): error TS2309: An export assignment cannot be used in a module with other exported elements.
 
 
-==== ambientErrors.ts (14 errors) ====
+==== ambientErrors.ts (15 errors) ====
     // Ambient variable with an initializer 
     declare var x = 4;
                     ~
 !!! error TS1039: Initializers are not allowed in ambient contexts.
@@ -88,8 +89,10 @@
     
     // Ambient enum with computer member
     declare enum E2 {
         x = 'foo'.length
+        ~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
             ~~~~~~~~~~~~
 !!! error TS1066: In ambient enum declarations member initializer must be constant expression.
     }
     
