// [[Reason: undefined]] ////

//// [tests/cases/compiler/classMemberWithMissingIdentifier.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -1,22 +1,25 @@
 
 
 //// [classMemberWithMissingIdentifier.d.ts]
 declare class C {
-    : any;
+    : invalid;
 }
 /// [Errors] ////
 
 classMemberWithMissingIdentifier.ts(2,11): error TS1146: Declaration expected.
+classMemberWithMissingIdentifier.ts(2,11): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 classMemberWithMissingIdentifier.ts(2,12): error TS1005: ';' expected.
 classMemberWithMissingIdentifier.ts(3,1): error TS1128: Declaration or statement expected.
 
 
-==== classMemberWithMissingIdentifier.ts (3 errors) ====
+==== classMemberWithMissingIdentifier.ts (4 errors) ====
     class C { 
         public {};
               
 !!! error TS1146: Declaration expected.
+              
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
                ~
 !!! error TS1005: ';' expected.
     }
     ~
