// [[Reason: undefined]] ////

//// [tests/cases/compiler/commonMissingSemicolons.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -11,9 +11,9 @@
 declare const myConst1 = 1;
 declare const myDeclareConst1: 1;
 declare const myDeclareConst2: 1;
 declare function myFunction1(): void;
-declare function (): any;
+declare function (): invalid;
 interface myInterface1 {
 }
 interface interface {
 }
@@ -78,8 +78,9 @@
 commonMissingSemicolons.ts(25,1): error TS1435: Unknown keyword or identifier. Did you mean 'function'?
 commonMissingSemicolons.ts(25,1): error TS2304: Cannot find name 'functiond'.
 commonMissingSemicolons.ts(25,11): error TS2304: Cannot find name 'myFunction2'.
 commonMissingSemicolons.ts(25,25): error TS1005: ';' expected.
+commonMissingSemicolons.ts(26,9): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 commonMissingSemicolons.ts(26,10): error TS1359: Identifier expected. 'function' is a reserved word that cannot be used here.
 commonMissingSemicolons.ts(26,18): error TS1003: Identifier expected.
 commonMissingSemicolons.ts(27,1): error TS2304: Cannot find name 'functionMyFunction'.
 commonMissingSemicolons.ts(30,1): error TS1435: Unknown keyword or identifier. Did you mean 'interface'?
@@ -121,9 +122,9 @@
 commonMissingSemicolons.ts(75,11): error TS1005: ';' expected.
 commonMissingSemicolons.ts(78,1): error TS1128: Declaration or statement expected.
 
 
-==== commonMissingSemicolons.ts (79 errors) ====
+==== commonMissingSemicolons.ts (80 errors) ====
     async function myAsyncFunction1(): Promise<void> {}
                                        ~~~~~~~~~~~~~
 !!! error TS2705: An async function or method in ES5/ES3 requires the 'Promise' constructor.  Make sure you have a declaration for the 'Promise' constructor or include 'ES2015' in your '--lib' option.
     asynd function myAsyncFunction2(): void {}
@@ -227,8 +228,10 @@
 !!! error TS2304: Cannot find name 'myFunction2'.
                             ~
 !!! error TS1005: ';' expected.
     function function(): void { }
+            
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
              ~~~~~~~~
 !!! error TS1359: Identifier expected. 'function' is a reserved word that cannot be used here.
                      ~
 !!! error TS1003: Identifier expected.
