// [[Reason: undefined]] ////

//// [tests/cases/compiler/constEnumNamespaceReferenceCausesNoImport2.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -12,7 +12,34 @@
 //// [index.d.ts]
 export {};
 
 //// [reexport.d.ts]
-import * as Foo from "./foo";
-declare const _default: typeof Foo.ConstEnumOnlyModule;
+declare const _default: invalid;
 export = _default;
+
 /// [Errors] ////
+
+reexport.ts(2,10): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+
+
+==== index.ts (0 errors) ====
+    import Foo = require("./reexport");
+    function check(x: Foo.ConstFooEnum): void {
+      switch (x) {
+        case Foo.ConstFooEnum.Some:
+          break;
+      }
+    }
+==== foo.ts (0 errors) ====
+    export module ConstEnumOnlyModule {
+      export const enum ConstFooEnum {
+        Some,
+        Values,
+        Here
+      }
+    }
+    
+==== reexport.ts (1 errors) ====
+    import * as Foo from "./foo";
+    export = Foo.ConstEnumOnlyModule;
+             ~~~~~~~~~~~~~~~~~~~~~~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+    
\ No newline at end of file
