// [[Reason: undefined]] ////

//// [tests/cases/compiler/constructorWithIncompleteTypeAnnotation.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -14,9 +14,9 @@
 declare namespace TypeScriptAllInOne {
     class Program {
         static Main(...args: string[]): void;
         case: any;
-        if(retValue: any): any;
+        if(retValue: any): invalid;
     }
 }
 declare class BasicFeatures {
     VARIABLES(): number;
@@ -65,8 +65,9 @@
 constructorWithIncompleteTypeAnnotation.ts(24,28): error TS1005: ':' expected.
 constructorWithIncompleteTypeAnnotation.ts(24,29): error TS1005: ',' expected.
 constructorWithIncompleteTypeAnnotation.ts(27,18): error TS1128: Declaration or statement expected.
 constructorWithIncompleteTypeAnnotation.ts(27,31): error TS2304: Cannot find name 'bfs'.
+constructorWithIncompleteTypeAnnotation.ts(28,17): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 constructorWithIncompleteTypeAnnotation.ts(28,35): error TS1005: ',' expected.
 constructorWithIncompleteTypeAnnotation.ts(28,39): error TS1005: ';' expected.
 constructorWithIncompleteTypeAnnotation.ts(31,18): error TS1109: Expression expected.
 constructorWithIncompleteTypeAnnotation.ts(34,17): error TS2304: Cannot find name 'retValue'.
@@ -150,9 +151,9 @@
 constructorWithIncompleteTypeAnnotation.ts(259,55): error TS1005: ';' expected.
 constructorWithIncompleteTypeAnnotation.ts(261,1): error TS1128: Declaration or statement expected.
 
 
-==== constructorWithIncompleteTypeAnnotation.ts (93 errors) ====
+==== constructorWithIncompleteTypeAnnotation.ts (94 errors) ====
     declare module "fs" {
         export class File {
             constructor(filename: string);
             public ReadAllText(): string;
@@ -203,8 +204,10 @@
 !!! error TS1128: Declaration or statement expected.
                                   ~~~
 !!! error TS2304: Cannot find name 'bfs'.
                     if (retValue: any != 0) {
+                    ~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
                                       ~~
 !!! error TS1005: ',' expected.
                                           ~
 !!! error TS1005: ';' expected.
