// [[Reason: undefined]] ////

//// [tests/cases/compiler/declFileEnums.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -28,4 +28,47 @@
     "Saturday" = 1,
     "Sunday" = 2,
     "Weekend days" = 3
 }
+/// [Errors] ////
+
+declFileEnums.ts(15,5): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+
+
+==== declFileEnums.ts (1 errors) ====
+    enum e1 {
+        a,
+        b,
+        c
+    }
+    
+    enum e2 {
+        a = 10,
+        b = a + 2,
+        c = 10,
+    }
+    
+    enum e3 {
+        a = 10,
+        b = Math.PI,
+        ~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+        c = a + 3
+    }
+    
+    enum e4 {
+        a,
+        b,
+        c,
+        d = 10,
+        e
+    }
+    
+    enum e5 {
+        "Friday",
+        "Saturday",
+        "Sunday",
+        "Weekend days"
+    }
+    
+    
+    
\ No newline at end of file
