// [[Reason: Syntactically invalid.]] ////

//// [tests/cases/compiler/declarationEmitDestructuringParameterProperties.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -1,29 +1,29 @@
 
 
 //// [declarationEmitDestructuringParameterProperties.d.ts]
 declare class C1 {
-    x: string;
-    y: string;
-    z: string;
+    x: invalid;
+    y: invalid;
+    z: invalid;
     constructor([x, y, z]: string[]);
 }
 type TupleType1 = [string, number, boolean];
 declare class C2 {
-    x: string;
-    y: number;
-    z: boolean;
+    x: invalid;
+    y: invalid;
+    z: invalid;
     constructor([x, y, z]: TupleType1);
 }
 type ObjType1 = {
     x: number;
     y: string;
     z: boolean;
 };
 declare class C3 {
-    x: number;
-    y: string;
-    z: boolean;
+    x: invalid;
+    y: invalid;
+    z: invalid;
     constructor({ x, y, z }: ObjType1);
 }
 //# sourceMappingURL=declarationEmitDestructuringParameterProperties.d.ts.map
 /// [Errors] ////
