// [[Reason: Printing differences]] ////

//// [tests/cases/compiler/declarationEmitInlinedDistributiveConditional.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -4,10 +4,14 @@
 export {};
 //# sourceMappingURL=test.d.ts.map
 //// [/.src/api.d.ts]
 import { PublicKeys1 } from './internal';
-export declare const dropPrivateProps1: <Obj>(obj: Obj) => { [K in PublicKeys1<keyof Obj>]: Obj[K]; };
-export declare const dropPrivateProps2: <Obj>(obj: Obj) => { [K in keyof Obj extends infer T ? T extends keyof Obj ? T extends `_${string}` ? never : T : never : never]: Obj[K]; };
+export declare const dropPrivateProps1: <Obj>(obj: Obj) => {
+    [K in PublicKeys1<keyof Obj>]: Obj[K];
+};
+export declare const dropPrivateProps2: <Obj>(obj: Obj) => {
+    [K in keyof Obj extends infer T ? T extends keyof Obj ? T extends `_${string}` ? never : T : never : never]: Obj[K];
+};
 //# sourceMappingURL=api.d.ts.map
 //// [/.src/internal.d.ts]
 export declare function excludePrivateKeys1<Obj>(obj: Obj): {
     [K in PublicKeys1<keyof Obj>]: Obj[K];
