// [[Reason: undefined]] ////

//// [tests/cases/conformance/es6/arrowFunction/disallowLineTerminatorBeforeArrow.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -27,14 +27,15 @@
     var v: any, any: any;
 }
 /// [Errors] ////
 
+disallowLineTerminatorBeforeArrow.ts(67,9): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 disallowLineTerminatorBeforeArrow.ts(71,25): error TS2304: Cannot find name 'x'.
 disallowLineTerminatorBeforeArrow.ts(71,26): error TS1005: ',' expected.
 disallowLineTerminatorBeforeArrow.ts(72,9): error TS1128: Declaration or statement expected.
 
 
-==== disallowLineTerminatorBeforeArrow.ts (3 errors) ====
+==== disallowLineTerminatorBeforeArrow.ts (4 errors) ====
     var f1 = (): void
         => { }
     var f2 = (x: string, y: string): void /*
       */  => { }
@@ -100,8 +101,10 @@
         }
     
         export enum Enum {
             claw = (()
+            ~~~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
                 => 10)()
         }
     
         export var v: any = x: any: any
