// [[Reason: undefined]] ////

//// [tests/cases/conformance/enums/enumErrorOnConstantBindingWithInitializer.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -13,13 +13,14 @@
 /// [Errors] ////
 
 enumErrorOnConstantBindingWithInitializer.ts(7,7): error TS2322: Type 'string | number | undefined' is not assignable to type 'string | number'.
   Type 'undefined' is not assignable to type 'string | number'.
+enumErrorOnConstantBindingWithInitializer.ts(10,3): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 enumErrorOnConstantBindingWithInitializer.ts(10,10): error TS18033: Type 'string | number' is not assignable to type 'number' as required for computed enum member values.
   Type 'string' is not assignable to type 'number'.
 
 
-==== enumErrorOnConstantBindingWithInitializer.ts (2 errors) ====
+==== enumErrorOnConstantBindingWithInitializer.ts (3 errors) ====
     type Thing = {
       value?: string | number;
     };
     
@@ -31,8 +32,10 @@
 !!! error TS2322:   Type 'undefined' is not assignable to type 'string | number'.
     
     enum E {
       test = value,
+      ~~~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
              ~~~~~
 !!! error TS18033: Type 'string | number' is not assignable to type 'number' as required for computed enum member values.
 !!! error TS18033:   Type 'string' is not assignable to type 'number'.
     }
