// [[Reason: undefined]] ////

//// [tests/cases/conformance/enums/enumMergingErrors.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -65,28 +65,37 @@
     }
 }
 
 /// [Errors] ////
 
+enumMergingErrors.ts(8,22): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+enumMergingErrors.ts(9,22): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+enumMergingErrors.ts(15,22): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 enumMergingErrors.ts(26,22): error TS2432: In an enum with multiple declarations, only one declaration can omit an initializer for its first enum element.
 enumMergingErrors.ts(38,22): error TS2432: In an enum with multiple declarations, only one declaration can omit an initializer for its first enum element.
 
 
-==== enumMergingErrors.ts (2 errors) ====
+==== enumMergingErrors.ts (5 errors) ====
     // Enum with constant, computed, constant members split across 3 declarations with the same root module
     module M {
         export enum E1 { A = 0 }
         export enum E2 { C }
         export enum E3 { A = 0 }
     }
     module M {
         export enum E1 { B = 'foo'.length }
+                         ~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
         export enum E2 { B = 'foo'.length }
+                         ~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
         export enum E3 { C }
     }
     module M {
         export enum E1 { C }
         export enum E2 { A = 0 }
         export enum E3 { B = 'foo'.length }
+                         ~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
     }
     
     // Enum with no initializer in either declaration with constant members with the same root module
     module M1 {
