// [[Reason: undefined]] ////

//// [tests/cases/compiler/enumWithExport.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -8,16 +8,19 @@
     z
 }
 /// [Errors] ////
 
+enumWithExport.ts(5,3): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 enumWithExport.ts(5,7): error TS2304: Cannot find name 'y'.
 
 
-==== enumWithExport.ts (1 errors) ====
+==== enumWithExport.ts (2 errors) ====
     namespace x {
       export let y = 123
     }
     enum x {
       z = y
+      ~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
           ~
 !!! error TS2304: Cannot find name 'y'.
     }
\ No newline at end of file
