// [[Reason: undefined]] ////

//// [tests/cases/compiler/enumWithoutInitializerAfterComputedMember.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -5,4 +5,17 @@
     a = 0,
     b = 0,
     c = 1
 }
+
 /// [Errors] ////
+
+enumWithoutInitializerAfterComputedMember.ts(3,5): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+
+
+==== enumWithoutInitializerAfterComputedMember.ts (1 errors) ====
+    enum E {
+        a,
+        b = a,
+        ~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+        c
+    }
\ No newline at end of file
