// [[Reason: undefined]] ////

//// [tests/cases/conformance/externalModules/exportAssignNonIdentifier.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -1,8 +1,8 @@
 
 
 //// [foo1.d.ts]
-declare const _default: "string" | "number" | "bigint" | "boolean" | "symbol" | "undefined" | "object" | "function";
+declare const _default: invalid;
 export = _default;
 
 //// [foo2.d.ts]
 declare const _default: "sausages";
@@ -21,26 +21,31 @@
 //// [foo5.d.ts]
 export = undefined;
 
 //// [foo6.d.ts]
-declare const _default: any;
+declare const _default: invalid;
 export = _default;
 
 //// [foo7.d.ts]
-declare const _default: DateConstructor | StringConstructor;
+declare const _default: invalid;
 export = _default;
 
 //// [foo8.d.ts]
 declare const _default: any;
 export = _default;
 /// [Errors] ////
 
+foo1.ts(2,10): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+foo6.ts(1,10): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 foo6.ts(1,14): error TS1109: Expression expected.
+foo7.ts(1,10): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 
 
-==== foo1.ts (0 errors) ====
+==== foo1.ts (1 errors) ====
     var x = 10;
     export = typeof x; // Ok
+             ~~~~~~~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
     
 ==== foo2.ts (0 errors) ====
     export = "sausages"; // Ok
     
@@ -52,15 +57,19 @@
     
 ==== foo5.ts (0 errors) ====
     export = undefined; // Valid.  undefined is an identifier in JavaScript/TypeScript
     
-==== foo6.ts (1 errors) ====
+==== foo6.ts (2 errors) ====
     export = void; // Error, void operator requires an argument
+             ~~~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
                  ~
 !!! error TS1109: Expression expected.
     
-==== foo7.ts (0 errors) ====
+==== foo7.ts (1 errors) ====
     export = Date || String; // Ok
+             ~~~~~~~~~~~~~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
     
 ==== foo8.ts (0 errors) ====
     export = null; // Ok
     
