// [[Reason: undefined]] ////

//// [tests/cases/compiler/exportEqualsProperty.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -1,21 +1,64 @@
 
 
 //// [a.d.ts]
-declare namespace A {
-    class B {
-        constructor(b: number);
-    }
-    namespace B {
-        const b: number;
-    }
-}
-declare const _default: typeof A.B;
+declare const _default: invalid;
 export = _default;
 
 //// [b.d.ts]
-declare const _default: number;
+declare const _default: invalid;
 export = _default;
 
 //// [index.d.ts]
-/// <reference path="declarations.d.ts" />
 export {};
+/// [Errors] ////
+
+a.ts(5,10): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+b.ts(1,10): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+index.ts(1,22): error TS9010: Reference directives are not supported in isolated declaration mode.
+
+
+==== index.ts (1 errors) ====
+    /// <reference path="declarations.d.ts" />
+                         ~~~~~~~~~~~~~~~~~
+!!! error TS9010: Reference directives are not supported in isolated declaration mode.
+    import { X } from "foobar";
+    import X2 = require("foobarx");
+    const x: X = X;
+    const x2: X2 = X2;
+    
+    import B = require("./a");
+    const b: B = new B(B.b);
+    
+    import fooLength = require("./b");
+    fooLength + 1;
+    
+==== declarations.d.ts (0 errors) ====
+    // This test is just like exportDefaultProperty, but with `export =`.
+    
+    declare namespace foo.bar {
+        export type X = number;
+        export const X: number;
+    }
+    
+    declare module "foobar" {
+        export = foo.bar;
+    }
+    
+    declare module "foobarx" {
+        export = foo.bar.X;
+    }
+    
+==== a.ts (1 errors) ====
+    namespace A {
+        export class B { constructor(b: number) {} }
+        export namespace B { export const b: number = 0; }
+    }
+    export = A.B;
+             ~~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+    
+==== b.ts (1 errors) ====
+    export = "foo".length;
+             ~~~~~~~~~~~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+    
\ No newline at end of file
