// [[Reason: undefined]] ////

//// [tests/cases/conformance/externalModules/exportAssignDottedName.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -3,7 +3,22 @@
 //// [foo1.d.ts]
 export declare function x(): boolean;
 
 //// [foo2.d.ts]
-import foo1 = require('./foo1');
-declare const _default: typeof foo1.x;
+declare const _default: invalid;
 export = _default;
+
 /// [Errors] ////
+
+foo2.ts(2,10): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+
+
+==== foo2.ts (1 errors) ====
+    import foo1 = require('./foo1');
+    export = foo1.x; // Ok
+             ~~~~~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+    
+==== foo1.ts (0 errors) ====
+    export function x(): boolean{
+    	return true;
+    }
+    
\ No newline at end of file
