// [[Reason: undefined]] ////

//// [tests/cases/conformance/functions/functionImplementations.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -13,25 +13,25 @@
 declare function rec4(): number;
 declare var n: number;
 declare var n: number;
 declare var n: number;
-declare var n: number;
+declare var n: invalid;
 declare var nu: any;
-declare var nu: any;
+declare var nu: invalid;
 declare var un: any;
-declare var un: any;
-declare var n: number;
-declare var n: number;
-declare var n: number;
+declare var un: invalid;
+declare var n: invalid;
+declare var n: invalid;
+declare var n: invalid;
 declare class Base {
     private m;
 }
 declare class Derived extends Base {
     private q;
 }
 declare var b: Base;
-declare var b: Base;
-declare var a: any;
+declare var b: invalid;
+declare var a: invalid;
 declare function thisFunc(): void;
 declare function opt1(n?: number): void;
 declare function opt2(n?: {
     x: any;
@@ -52,14 +52,22 @@
 declare var f11: (x: number) => any;
 declare var f12: (x: number) => any;
 /// [Errors] ////
 
+functionImplementations.ts(40,9): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+functionImplementations.ts(46,10): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+functionImplementations.ts(52,10): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+functionImplementations.ts(57,9): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+functionImplementations.ts(62,9): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 functionImplementations.ts(67,5): error TS2403: Subsequent variable declarations must have the same type.  Variable 'n' must be of type 'number', but here has type '3 | 5'.
+functionImplementations.ts(67,9): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+functionImplementations.ts(80,9): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 functionImplementations.ts(85,5): error TS2403: Subsequent variable declarations must have the same type.  Variable 'a' must be of type 'any', but here has type 'Base'.
+functionImplementations.ts(85,9): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 functionImplementations.ts(90,1): error TS2839: This condition will always return 'false' since JavaScript compares objects by reference, not value.
 
 
-==== functionImplementations.ts (3 errors) ====
+==== functionImplementations.ts (11 errors) ====
     // FunctionExpression with no return type annotation and no return statement returns void
     var v: void = function () { } ();
     
     // FunctionExpression f with no return type annotation and directly references f in its body returns any
@@ -98,41 +106,66 @@
     var n: number = rec4();
     
     // FunctionExpression with no return type annotation and returns a number
     var n = function (): number {
+            ~~~~~~~~~~~~~~~~~~~~~
         return 3;
+    ~~~~~~~~~~~~~
     } ();
+    ~~~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
     
     // FunctionExpression with no return type annotation and returns null
     var nu = null;
     var nu = function (): any {
+             ~~~~~~~~~~~~~~~~~~
         return null;
+    ~~~~~~~~~~~~~~~~
     } ();
+    ~~~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
     
     // FunctionExpression with no return type annotation and returns undefined
     var un = undefined;
     var un = function (): any {
+             ~~~~~~~~~~~~~~~~~~
         return undefined;
+    ~~~~~~~~~~~~~~~~~~~~~
     } ();
+    ~~~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
     
     // FunctionExpression with no return type annotation and returns a type parameter type
     var n = function <T>(x: T): T {
+            ~~~~~~~~~~~~~~~~~~~~~~~
         return x;
+    ~~~~~~~~~~~~~
     } (4);
+    ~~~~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
     
     // FunctionExpression with no return type annotation and returns a constrained type parameter type
     var n = function <T extends {}>(x: T): T {
+            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         return x;
+    ~~~~~~~~~~~~~
     } (4);
+    ~~~~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
     
     // FunctionExpression with no return type annotation with multiple return statements with identical types
     var n = function (): 3 | 5 {
         ~
 !!! error TS2403: Subsequent variable declarations must have the same type.  Variable 'n' must be of type 'number', but here has type '3 | 5'.
 !!! related TS6203 functionImplementations.ts:35:5: 'n' was also declared here.
+            ~~~~~~~~~~~~~~~~~~~~
         return 3;
+    ~~~~~~~~~~~~~
         return 5;
+    ~~~~~~~~~~~~~
     }();
+    ~~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
     
     // Otherwise, the inferred return type is the first of the types of the return statement expressions
     // in the function body that is a supertype of each of the others, 
     // ignoring return statements with no expressions.
@@ -141,18 +174,26 @@
     class Base { private m; }
     class Derived extends Base { private q; }
     var b: Base;
     var b = function (): Base {
+            ~~~~~~~~~~~~~~~~~~~
         return new Base(); return new Derived();
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     } ();
+    ~~~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
     
     // FunctionExpression with no return type annotation with multiple return statements with one a recursive call
     var a = function f(): Base {
         ~
 !!! error TS2403: Subsequent variable declarations must have the same type.  Variable 'a' must be of type 'any', but here has type 'Base'.
 !!! related TS6203 functionImplementations.ts:5:5: 'a' was also declared here.
+            ~~~~~~~~~~~~~~~~~~~~
         return new Base(); return new Derived(); return f(); // ?
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     } ();
+    ~~~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
     
     // FunctionExpression with non -void return type annotation with a single throw statement
     undefined === function (): number {
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
