// [[Reason: undefined]] ////

//// [tests/cases/conformance/salsa/jsContainerMergeTsDeclaration.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -1,13 +1,14 @@
 
 
 //// [b.d.ts]
-declare var x: number;
+declare var x: invalid;
 
 /// [Errors] ////
 
 error TS6504: File 'a.js' is a JavaScript file. Did you mean to enable the 'allowJs' option?
   The file is in the program because:
     Root file specified for compilation
+b.ts(1,9): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 
 
 !!! error TS6504: File 'a.js' is a JavaScript file. Did you mean to enable the 'allowJs' option?
 !!! error TS6504:   The file is in the program because:
@@ -16,9 +17,13 @@
     var /*1*/x = function foo() {
     }
     x.a = function bar() {
     }
-==== b.ts (0 errors) ====
+==== b.ts (1 errors) ====
     var x = function (): number {
+            ~~~~~~~~~~~~~~~~~~~~~
         return 1;
+    ~~~~~~~~~~~~~
     }();
+    ~~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
     
\ No newline at end of file
