// [[Reason: undefined]] ////

//// [tests/cases/conformance/parser/ecmascript5/parserRealSource14.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -84,8 +84,9 @@
 }
 
 /// [Errors] ////
 
 parserRealSource14.ts(4,21): error TS6053: File 'typescript.ts' not found.
+parserRealSource14.ts(4,21): error TS9010: Reference directives are not supported in isolated declaration mode.
 parserRealSource14.ts(24,33): error TS2694: Namespace 'TypeScript' has no exported member 'AST'.
 parserRealSource14.ts(38,34): error TS2694: Namespace 'TypeScript' has no exported member 'AST'.
 parserRealSource14.ts(48,37): error TS2694: Namespace 'TypeScript' has no exported member 'AST'.
 parserRealSource14.ts(68,39): error TS2694: Namespace 'TypeScript' has no exported member 'NodeType'.
@@ -221,8 +222,9 @@
 parserRealSource14.ts(422,30): error TS2694: Namespace 'TypeScript' has no exported member 'CallExpression'.
 parserRealSource14.ts(427,65): error TS2339: Property 'NodeType' does not exist on type 'typeof TypeScript'.
 parserRealSource14.ts(428,30): error TS2694: Namespace 'TypeScript' has no exported member 'Block'.
 parserRealSource14.ts(432,52): error TS2694: Namespace 'TypeScript' has no exported member 'ASTSpan'.
+parserRealSource14.ts(456,9): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 parserRealSource14.ts(462,61): error TS2694: Namespace 'TypeScript' has no exported member 'AST'.
 parserRealSource14.ts(463,52): error TS2694: Namespace 'TypeScript' has no exported member 'Comment'.
 parserRealSource14.ts(478,45): error TS2694: Namespace 'TypeScript' has no exported member 'AST'.
 parserRealSource14.ts(478,69): error TS2694: Namespace 'TypeScript' has no exported member 'AST'.
@@ -247,15 +249,17 @@
 parserRealSource14.ts(565,94): error TS2694: Namespace 'TypeScript' has no exported member 'IAstWalker'.
 parserRealSource14.ts(572,20): error TS2339: Property 'getAstWalkerFactory' does not exist on type 'typeof TypeScript'.
 
 
-==== parserRealSource14.ts (162 errors) ====
+==== parserRealSource14.ts (164 errors) ====
     // Copyright (c) Microsoft. All rights reserved. Licensed under the Apache License, Version 2.0. 
     // See LICENSE.txt in the project root for complete license information.
     
     ///<reference path='typescript.ts' />
                         ~~~~~~~~~~~~~
 !!! error TS6053: File 'typescript.ts' not found.
+                        ~~~~~~~~~~~~~
+!!! error TS9010: Reference directives are not supported in isolated declaration mode.
     
     module TypeScript {
         export function lastOf(items: any[]): any {
             return (items === null || items.length === 0) ? null : items[items.length - 1];
@@ -980,8 +984,10 @@
             // the "{" character, meaning we don't traverse the tree down to the stmt list of the class, meaning
             // we don't find the "precomment" attached to the errorneous empty stmt.
             //TODO: It would be nice to be able to get rid of this.
             DontPruneSearchBasedOnPosition = 1 << 1,
+            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
         }
     
         ///
         /// Return the stack of AST nodes containing "position"
