// [[Reason: undefined]] ////

//// [tests/cases/compiler/staticsInAFunction.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -1,10 +1,11 @@
 
 
 //// [/.src/staticsInAFunction.d.ts]
-declare function boo(): void;
+declare function boo(): invalid;
 /// [Errors] ////
 
+staticsInAFunction.ts(1,10): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 staticsInAFunction.ts(1,13): error TS1005: '(' expected.
 staticsInAFunction.ts(2,4): error TS1128: Declaration or statement expected.
 staticsInAFunction.ts(2,11): error TS2582: Cannot find name 'test'. Do you need to install type definitions for a test runner? Try `npm i --save-dev @types/jest` or `npm i --save-dev @types/mocha`.
 staticsInAFunction.ts(3,4): error TS1128: Declaration or statement expected.
@@ -19,10 +20,12 @@
 staticsInAFunction.ts(4,22): error TS2693: 'any' only refers to a type, but is being used as a value here.
 staticsInAFunction.ts(4,26): error TS1005: ';' expected.
 
 
-==== staticsInAFunction.ts (14 errors) ====
+==== staticsInAFunction.ts (15 errors) ====
     function boo{
+             ~~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
                 ~
 !!! error TS1005: '(' expected.
        static test()
        ~~~~~~
