// [[Reason: undefined]] ////

//// [tests/cases/conformance/expressions/literals/strictModeOctalLiterals.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -5,16 +5,19 @@
     A = 13
 }
 /// [Errors] ////
 
+strictModeOctalLiterals.ts(2,5): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 strictModeOctalLiterals.ts(2,14): error TS1121: Octal literals are not allowed. Use the syntax '0o1'.
 strictModeOctalLiterals.ts(4,16): error TS1121: Octal literals are not allowed. Use the syntax '0o1'.
 strictModeOctalLiterals.ts(4,21): error TS1121: Octal literals are not allowed. Use the syntax '0o1'.
 
 
-==== strictModeOctalLiterals.ts (3 errors) ====
+==== strictModeOctalLiterals.ts (4 errors) ====
     export enum E {
         A = 12 + 01
+        ~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
                  ~~
 !!! error TS1121: Octal literals are not allowed. Use the syntax '0o1'.
     }
     const orbitol: 01 = 01
