// [[Reason: undefined]] ////

//// [tests/cases/conformance/expressions/thisKeyword/thisInInvalidContexts.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -28,13 +28,15 @@
 thisInInvalidContexts.ts(17,15): error TS17009: 'super' must be called before accessing 'this' in the constructor of a derived class.
 thisInInvalidContexts.ts(23,13): error TS2331: 'this' cannot be referenced in a module or namespace body.
 thisInInvalidContexts.ts(31,13): error TS2526: A 'this' type is available only in a non-static member of a class or interface.
 thisInInvalidContexts.ts(34,25): error TS2507: Type 'typeof globalThis' is not a constructor function type.
+thisInInvalidContexts.ts(40,5): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 thisInInvalidContexts.ts(40,9): error TS2332: 'this' cannot be referenced in current location.
+thisInInvalidContexts.ts(41,5): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 thisInInvalidContexts.ts(41,9): error TS2332: 'this' cannot be referenced in current location.
 
 
-==== thisInInvalidContexts.ts (7 errors) ====
+==== thisInInvalidContexts.ts (9 errors) ====
     class BaseErrClass {
         constructor(t: any) { }
     }
     
@@ -83,11 +85,15 @@
     
     //'this' as a computed enum value
     enum SomeEnum {
         A = this, // Should not be allowed
+        ~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
             ~~~~
 !!! error TS2332: 'this' cannot be referenced in current location.
         B = this.spaaaace // Also should not be allowed
+        ~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
             ~~~~
 !!! error TS2332: 'this' cannot be referenced in current location.
     }
     
