// [[Reason: undefined]] ////

//// [tests/cases/compiler/thisInPropertyBoundDeclarations.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -24,16 +24,17 @@
     prop4: string;
     prop5: {
         a: () => this;
     };
-    prop6: any;
+    prop6: invalid;
 }
 /// [Errors] ////
 
 thisInPropertyBoundDeclarations.ts(64,5): error TS2527: The inferred type of 'prop6' references an inaccessible 'this' type. A type annotation is necessary.
+thisInPropertyBoundDeclarations.ts(64,13): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 
 
-==== thisInPropertyBoundDeclarations.ts (1 errors) ====
+==== thisInPropertyBoundDeclarations.ts (2 errors) ====
     class Bug {
         private name: string;
     
         private static func: Function[] = [
@@ -98,8 +99,10 @@
     
         prop6 = () => {
         ~~~~~
 !!! error TS2527: The inferred type of 'prop6' references an inaccessible 'this' type. A type annotation is necessary.
+                ~~~~~~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
             return {
                 a: () => { return this; }
             };
         };
