// [[Reason: Syntactically invalid.]] ////

//// [tests/cases/compiler/accessorBodyInTypeContext.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -1,15 +1,15 @@
 
 
 //// [accessorBodyInTypeContext.d.ts]
 type A = {
-    get foo(): number;
+    get foo(): any;
 };
 type B = {
     set foo(v: any);
 };
 interface X {
-    get foo(): number;
+    get foo(): any;
 }
 interface Y {
     set foo(v: any);
 }
