// [[Reason: TS normalizes types]] ////

//// [tests/cases/compiler/booleanFilterAnyArray.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -21,13 +21,13 @@
 declare var foo: invalid;
 declare var foor: Array<{
     name: string;
 }>;
-declare var foor: {
+declare var foor: Array<{
     name: string;
-}[];
+}>;
 declare var foos: Array<boolean>;
-declare var foos: boolean[];
+declare var foos: Array<boolean>;
 
 /// [Errors] ////
 
 booleanFilterAnyArray.ts(20,11): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
