// [[Reason: undefined]] ////

//// [tests/cases/compiler/constEnums.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -48,9 +48,9 @@
     namespace B {
         namespace C {
             const enum E {
                 V1 = 1,
-                V2 = 101
+                V2
             }
         }
     }
 }
@@ -102,17 +102,18 @@
 constEnums.ts(34,5): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 constEnums.ts(34,10): error TS2474: const enum member initializers must be constant expressions.
 constEnums.ts(35,5): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 constEnums.ts(35,10): error TS2474: const enum member initializers must be constant expressions.
+constEnums.ts(55,17): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 constEnums.ts(65,17): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 constEnums.ts(65,22): error TS2474: const enum member initializers must be constant expressions.
 constEnums.ts(66,17): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 constEnums.ts(66,22): error TS2474: const enum member initializers must be constant expressions.
 constEnums.ts(124,10): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 constEnums.ts(166,10): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 
 
-==== constEnums.ts (12 errors) ====
+==== constEnums.ts (13 errors) ====
     const enum Enum1 {
        A0 = 100,
     }
     
@@ -178,8 +179,10 @@
             export module C {
                 export const enum E {
                     V1 = 1,
                     V2 = A.B.C.E.V1 | 100
+                    ~~
+!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
                 }
             }
         }
     }
