// [[Reason: undefined]] ////

//// [tests/cases/compiler/declarationEmitForModuleImportingModuleAugmentationRetainsImport.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -14,9 +14,8 @@
 }
 
 /// [Errors] ////
 
 child1.ts(9,17): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
-parent.ts(1,1): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 
 
 ==== child1.ts (1 errors) ====
     import { ParentThing } from './parent';
@@ -32,12 +31,10 @@
 !!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
         prototype.add = (a: number, b: number) => a + b;
     }
     
-==== parent.ts (1 errors) ====
+==== parent.ts (0 errors) ====
     import { child1 } from './child1'; // this import should still exist in some form in the output, since it augments this module
-    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
     
     export class ParentThing implements ParentThing {}
     
     child1(ParentThing.prototype);
\ No newline at end of file
