// [[Reason: undefined]] ////

//// [tests/cases/compiler/declarationEmitLateBoundAssignments2.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -44,24 +44,29 @@
 declarationEmitLateBoundAssignments2.ts(36,17): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 declarationEmitLateBoundAssignments2.ts(36,17): error TS9009: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
 declarationEmitLateBoundAssignments2.ts(39,14): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 declarationEmitLateBoundAssignments2.ts(39,14): error TS9009: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
-declarationEmitLateBoundAssignments2.ts(42,23): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+declarationEmitLateBoundAssignments2.ts(42,14): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+declarationEmitLateBoundAssignments2.ts(42,14): error TS9009: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
 declarationEmitLateBoundAssignments2.ts(45,14): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 declarationEmitLateBoundAssignments2.ts(45,14): error TS9009: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
-declarationEmitLateBoundAssignments2.ts(48,23): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+declarationEmitLateBoundAssignments2.ts(48,14): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+declarationEmitLateBoundAssignments2.ts(48,14): error TS9009: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
 declarationEmitLateBoundAssignments2.ts(51,14): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 declarationEmitLateBoundAssignments2.ts(51,14): error TS9009: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
-declarationEmitLateBoundAssignments2.ts(54,23): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+declarationEmitLateBoundAssignments2.ts(54,14): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+declarationEmitLateBoundAssignments2.ts(54,14): error TS9009: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
 declarationEmitLateBoundAssignments2.ts(57,14): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 declarationEmitLateBoundAssignments2.ts(57,14): error TS9009: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
-declarationEmitLateBoundAssignments2.ts(60,23): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+declarationEmitLateBoundAssignments2.ts(60,14): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+declarationEmitLateBoundAssignments2.ts(60,14): error TS9009: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
 declarationEmitLateBoundAssignments2.ts(63,14): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 declarationEmitLateBoundAssignments2.ts(63,14): error TS9009: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
-declarationEmitLateBoundAssignments2.ts(66,24): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+declarationEmitLateBoundAssignments2.ts(66,14): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+declarationEmitLateBoundAssignments2.ts(66,14): error TS9009: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
 
 
-==== declarationEmitLateBoundAssignments2.ts (35 errors) ====
+==== declarationEmitLateBoundAssignments2.ts (40 errors) ====
     // https://github.com/microsoft/TypeScript/issues/54811
     
     const c = "C"
     const num = 1
@@ -146,10 +151,12 @@
 !!! error TS9009: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
     arrow["B"] = 'bar'
     
     export const arrow2 = () => {}
-                          ~~~~~~~~
+                 ~~~~~~
 !!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+                 ~~~~~~
+!!! error TS9009: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
     arrow2[c] = 100
     
     export const arrow3 = () => {}
                  ~~~~~~
@@ -158,10 +165,12 @@
 !!! error TS9009: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
     arrow3[77] = 0
     
     export const arrow4 = () => {}
-                          ~~~~~~~~
+                 ~~~~~~
 !!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+                 ~~~~~~
+!!! error TS9009: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
     arrow4[num] = 0
     
     export const arrow5 = () => {}
                  ~~~~~~
@@ -170,10 +179,12 @@
 !!! error TS9009: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
     arrow5["101"] = 0
     
     export const arrow6 = () => {}
-                          ~~~~~~~~
+                 ~~~~~~
 !!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+                 ~~~~~~
+!!! error TS9009: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
     arrow6[numStr] = 0
     
     export const arrow7 = () => {}
                  ~~~~~~
@@ -182,10 +193,12 @@
 !!! error TS9009: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
     arrow7["qwe rty"] = 0
     
     export const arrow8 = () => {}
-                          ~~~~~~~~
+                 ~~~~~~
 !!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+                 ~~~~~~
+!!! error TS9009: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
     arrow8[withWhitespace] = 0
     
     export const arrow9 = () => {}
                  ~~~~~~
@@ -194,8 +207,10 @@
 !!! error TS9009: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
     arrow9["🤪"] = 0
     
     export const arrow10 = () => {}
-                           ~~~~~~~~
+                 ~~~~~~~
 !!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
+                 ~~~~~~~
+!!! error TS9009: Assigning properties to functions without declaring them is not supported with --isolatedDeclarations. Add an explicit declaration for the properties assigned to this function.
     arrow10[emoji] = 0
     
\ No newline at end of file
