// [[Reason: undefined]] ////

//// [tests/cases/compiler/decoratorsOnComputedProperties.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -79,13 +79,10 @@
 decoratorsOnComputedProperties.ts(18,5): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
 decoratorsOnComputedProperties.ts(19,8): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
 decoratorsOnComputedProperties.ts(20,8): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
 decoratorsOnComputedProperties.ts(21,5): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-decoratorsOnComputedProperties.ts(21,5): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 decoratorsOnComputedProperties.ts(22,8): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-decoratorsOnComputedProperties.ts(22,8): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 decoratorsOnComputedProperties.ts(23,8): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-decoratorsOnComputedProperties.ts(23,8): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 decoratorsOnComputedProperties.ts(27,5): error TS1206: Decorators are not valid here.
 decoratorsOnComputedProperties.ts(28,5): error TS1206: Decorators are not valid here.
 decoratorsOnComputedProperties.ts(29,5): error TS1206: Decorators are not valid here.
 decoratorsOnComputedProperties.ts(30,5): error TS1206: Decorators are not valid here.
@@ -98,13 +95,10 @@
 decoratorsOnComputedProperties.ts(52,5): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
 decoratorsOnComputedProperties.ts(53,8): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
 decoratorsOnComputedProperties.ts(54,8): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
 decoratorsOnComputedProperties.ts(55,5): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-decoratorsOnComputedProperties.ts(55,5): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 decoratorsOnComputedProperties.ts(56,8): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-decoratorsOnComputedProperties.ts(56,8): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 decoratorsOnComputedProperties.ts(57,8): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-decoratorsOnComputedProperties.ts(57,8): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 decoratorsOnComputedProperties.ts(62,5): error TS1206: Decorators are not valid here.
 decoratorsOnComputedProperties.ts(63,5): error TS1206: Decorators are not valid here.
 decoratorsOnComputedProperties.ts(64,5): error TS1206: Decorators are not valid here.
 decoratorsOnComputedProperties.ts(65,5): error TS1206: Decorators are not valid here.
@@ -117,13 +111,10 @@
 decoratorsOnComputedProperties.ts(88,5): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
 decoratorsOnComputedProperties.ts(89,8): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
 decoratorsOnComputedProperties.ts(90,8): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
 decoratorsOnComputedProperties.ts(92,5): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-decoratorsOnComputedProperties.ts(92,5): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 decoratorsOnComputedProperties.ts(93,8): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-decoratorsOnComputedProperties.ts(93,8): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 decoratorsOnComputedProperties.ts(94,8): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-decoratorsOnComputedProperties.ts(94,8): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 decoratorsOnComputedProperties.ts(98,5): error TS1206: Decorators are not valid here.
 decoratorsOnComputedProperties.ts(99,5): error TS1206: Decorators are not valid here.
 decoratorsOnComputedProperties.ts(100,5): error TS1206: Decorators are not valid here.
 decoratorsOnComputedProperties.ts(101,5): error TS1206: Decorators are not valid here.
@@ -136,13 +127,10 @@
 decoratorsOnComputedProperties.ts(124,5): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
 decoratorsOnComputedProperties.ts(125,8): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
 decoratorsOnComputedProperties.ts(126,8): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
 decoratorsOnComputedProperties.ts(128,5): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-decoratorsOnComputedProperties.ts(128,5): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 decoratorsOnComputedProperties.ts(129,8): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-decoratorsOnComputedProperties.ts(129,8): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 decoratorsOnComputedProperties.ts(131,8): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-decoratorsOnComputedProperties.ts(131,8): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 decoratorsOnComputedProperties.ts(135,5): error TS1206: Decorators are not valid here.
 decoratorsOnComputedProperties.ts(136,5): error TS1206: Decorators are not valid here.
 decoratorsOnComputedProperties.ts(137,5): error TS1206: Decorators are not valid here.
 decoratorsOnComputedProperties.ts(138,5): error TS1206: Decorators are not valid here.
@@ -155,13 +143,10 @@
 decoratorsOnComputedProperties.ts(162,5): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
 decoratorsOnComputedProperties.ts(163,8): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
 decoratorsOnComputedProperties.ts(164,8): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
 decoratorsOnComputedProperties.ts(166,5): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-decoratorsOnComputedProperties.ts(166,5): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 decoratorsOnComputedProperties.ts(167,8): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-decoratorsOnComputedProperties.ts(167,8): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 decoratorsOnComputedProperties.ts(169,8): error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-decoratorsOnComputedProperties.ts(169,8): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 decoratorsOnComputedProperties.ts(173,5): error TS1206: Decorators are not valid here.
 decoratorsOnComputedProperties.ts(174,5): error TS1206: Decorators are not valid here.
 decoratorsOnComputedProperties.ts(175,5): error TS1206: Decorators are not valid here.
 decoratorsOnComputedProperties.ts(176,5): error TS1206: Decorators are not valid here.
@@ -173,9 +158,9 @@
 decoratorsOnComputedProperties.ts(186,5): error TS1206: Decorators are not valid here.
 decoratorsOnComputedProperties.ts(188,5): error TS1206: Decorators are not valid here.
 
 
-==== decoratorsOnComputedProperties.ts (97 errors) ====
+==== decoratorsOnComputedProperties.ts (82 errors) ====
     function x(o: object, k: PropertyKey) { }
              ~
 !!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
     let i = 0;
@@ -205,20 +190,14 @@
 !!! error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
         [fieldNameA]: any;
         ~~~~~~~~~~~~
 !!! error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-        ~~~~~~~~~~~~
-!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
         @x [fieldNameB]: any;
            ~~~~~~~~~~~~
 !!! error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-           ~~~~~~~~~~~~
-!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
         @x [fieldNameC]: any = null;
            ~~~~~~~~~~~~
 !!! error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-           ~~~~~~~~~~~~
-!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
     }
     
     void class B {
         @x ["property"]: any;
@@ -277,20 +256,14 @@
 !!! error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
         [fieldNameA]: any;
         ~~~~~~~~~~~~
 !!! error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-        ~~~~~~~~~~~~
-!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
         @x [fieldNameB]: any;
            ~~~~~~~~~~~~
 !!! error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-           ~~~~~~~~~~~~
-!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
         @x [fieldNameC]: any = null;
            ~~~~~~~~~~~~
 !!! error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-           ~~~~~~~~~~~~
-!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
         ["some" + "method"]() {}
     }
     
     void class D {
@@ -352,20 +325,14 @@
         ["some" + "method"]() {}
         [fieldNameA]: any;
         ~~~~~~~~~~~~
 !!! error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-        ~~~~~~~~~~~~
-!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
         @x [fieldNameB]: any;
            ~~~~~~~~~~~~
 !!! error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-           ~~~~~~~~~~~~
-!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
         @x [fieldNameC]: any = null;
            ~~~~~~~~~~~~
 !!! error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-           ~~~~~~~~~~~~
-!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
     }
     
     void class F {
         @x ["property"]: any;
@@ -426,21 +393,15 @@
         ["some" + "method"]() {}
         [fieldNameA]: any;
         ~~~~~~~~~~~~
 !!! error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-        ~~~~~~~~~~~~
-!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
         @x [fieldNameB]: any;
            ~~~~~~~~~~~~
 !!! error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-           ~~~~~~~~~~~~
-!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
         ["some" + "method2"]() {}
         @x [fieldNameC]: any = null;
            ~~~~~~~~~~~~
 !!! error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-           ~~~~~~~~~~~~
-!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
     }
     
     void class H {
         @x ["property"]: any;
@@ -502,21 +463,15 @@
         @x ["some" + "method"]() {}
         [fieldNameA]: any;
         ~~~~~~~~~~~~
 !!! error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-        ~~~~~~~~~~~~
-!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
         @x [fieldNameB]: any;
            ~~~~~~~~~~~~
 !!! error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-           ~~~~~~~~~~~~
-!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
         ["some" + "method2"]() {}
         @x [fieldNameC]: any = null;
            ~~~~~~~~~~~~
 !!! error TS1166: A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.
-           ~~~~~~~~~~~~
-!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
     }
     
     void class J {
         @x ["property"]: any;
