// [[Reason: undefined]] ////

//// [tests/cases/conformance/enums/enumConstantMembers.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -22,20 +22,20 @@
     a = Infinity,
     b = Infinity,
     c = Infinity,
     d = NaN,
-    e = NaN,
-    f = Infinity,
-    g = -Infinity
+    e,
+    f,
+    g
 }
 declare const enum E6 {
     a = Infinity,
     b = Infinity,
     c = Infinity,
     d = NaN,
-    e = NaN,
-    f = Infinity,
-    g = -Infinity
+    e,
+    f,
+    g
 }
 /// [Errors] ////
 
 enumConstantMembers.ts(22,5): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
