// [[Reason: undefined]] ////

//// [tests/cases/compiler/indexSignatureMustHaveTypeAnnotation.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -1,8 +1,9 @@
 
 
 //// [/.src/indexSignatureMustHaveTypeAnnotation.d.ts]
 interface I {
+    [x]: string;
     [x: string]: any;
 }
 declare class C {
     [x]: string;
