// [[Reason: undefined]] ////

//// [tests/cases/compiler/isolatedModulesGlobalNamespacesAndEnums.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -12,15 +12,15 @@
 declare const d = "d";
 
 //// [/.src/enum2.d.ts]
 declare enum Enum {
-    D = "d",
-    E = 0,// error
-    Y = 1000000,// error
-    Z = 0
+    D,
+    E,// error
+    Y,// error
+    Z
 }
 declare enum Enum {
-    F = 0
+    F
 }
 
 //// [/.src/module-namespaces.d.ts]
 export declare namespace Instantiated {
