// [[Reason: undefined]] ////

//// [tests/cases/compiler/propertyAssignment.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -6,9 +6,11 @@
 };
 declare var bar1: {
     x: number;
 };
-declare var foo2: {};
+declare var foo2: {
+    [index]: any;
+};
 declare var bar2: {
     x: number;
 };
 declare var foo3: {
