// [[Reason: undefined]] ////

//// [tests/cases/conformance/es6/Symbols/symbolProperty58.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -3,20 +3,7 @@
 //// [symbolProperty58.d.ts]
 interface SymbolConstructor {
     foo: string;
 }
-declare var obj: invalid;
-/// [Errors] ////
-
-symbolProperty58.ts(6,5): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
-
-
-==== symbolProperty58.ts (1 errors) ====
-    interface SymbolConstructor {
-        foo: string;
-    }
-    
-    var obj = {
-        [Symbol.foo]: 0
-        ~~~~~~~~~~~~
-!!! error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
-    }
\ No newline at end of file
+declare var obj: {
+    [Symbol.foo]: number;
+};
