// [[Reason: undefined]] ////

//// [tests/cases/conformance/es6/Symbols/symbolDeclarationEmit12.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -5,8 +5,9 @@
     interface I {
     }
     export class C {
         [Symbol.iterator]: I;
+        [Symbol.toPrimitive](x: I): invalid;
         [Symbol.isConcatSpreadable](): I;
         get [Symbol.toPrimitive](): invalid;
         set [Symbol.toPrimitive](x: I);
     }
