// [[Reason: undefined]] ////

//// [tests/cases/conformance/es6/Symbols/symbolProperty59.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -1,8 +1,9 @@
 
 
 //// [symbolProperty59.d.ts]
 interface I {
+    [Symbol.keyFor]: string;
 }
 /// [Errors] ////
 
 symbolProperty59.ts(2,5): error TS1169: A computed property name in an interface must refer to an expression whose type is a literal type or a 'unique symbol' type.
