// [[Reason: undefined]] ////

//// [tests/cases/conformance/types/literal/templateLiteralTypes4.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -38,10 +38,10 @@
     One = 1
 }
 type TNumberLiteralEnum0 = "0" extends `${infer T extends NumberLiteralEnum}` ? T : never;
 declare const enum NonLiteralEnum {
-    Zero = 0,
-    One = 1
+    Zero,
+    One
 }
 type TNonLiteralEnum0 = "0" extends `${infer T extends NonLiteralEnum}` ? T : never;
 type PString00 = "0" extends `${infer T extends string | StringLiteralEnum}` ? T : never;
 type PString01 = "0" extends `${infer T extends string | number}` ? T : never;
