// [[Reason: undefined]] ////

//// [tests/cases/compiler/typeReferenceDirectives11.ts] ////

===================================================================
--- TSC	declarations
+++ DTE	declarations
@@ -7,9 +7,8 @@
 export declare const bar: invalid;
 
 /// [Errors] ////
 
-/mod1.ts(1,24): error TS9008: Declaration emit for this file requires adding a type reference directive. Add a type reference directive to lib to unblock declaration emit.
 /mod2.ts(2,20): error TS9007: Declaration emit for this file requires type resolution. An explicit type annotation may unblock declaration emit.
 
 
 ==== /mod2.ts (1 errors) ====
@@ -20,9 +19,7 @@
     
 ==== /types/lib/index.d.ts (0 errors) ====
     interface Lib { x }
     
-==== /mod1.ts (1 errors) ====
+==== /mod1.ts (0 errors) ====
     export function foo(): Lib { return {x: 1} }
-                           ~~~
-!!! error TS9008: Declaration emit for this file requires adding a type reference directive. Add a type reference directive to lib to unblock declaration emit.
     
\ No newline at end of file
